### ISSUE TITLE
AOTI compilation produces corrupt results for model.

### ISSUE THREAD

**Issue by mikekgfb** [pytorch/pytorch#122652]
### 🐛 Describe the bug

Exporting and aoti-compiling the model used in gpt-fast causes incorrect execution. 

repro:
Download stories15M from HF
(internally, use `jf download GICWmADZB8jycYwBAJw9dzU9F7wObpZBAAAB --file "checkpoints.tgz"; tar xvzf checkpoints.tgz`)

```
# works, normal stories15M.pt model with gpt-fast
python generate.py --checkpoint_path checkpoints/$MODEL_REPO/stories15M.pt --prompt "Once upon a time" --device cpu --temperature 0 

# compile model into dso
python export.py --checkpoint_path checkpoints/$MODEL_REPO/stories15M.pt --prompt "Once upon a time" --device cpu --temperature 0 
cp $so_file stories15M.so

# use the dso
python generate.py --checkpoint_path checkpoints/$MODEL_REPO/stories15M.pt --prompt "Once upon a time" --device cpu --temperature 0 --dso
```

Download from https://github.com/mikekgfb/llama-fast or use these files:

[export.py.txt](https://github.com/pytorch/pytorch/files/14750197/export.py.txt)
[generate.py.txt](https://github.com/pytorch/pytorch/files/14750198/generate.py.txt)
[model.py.txt](https://github.com/pytorch/pytorch/files/14750199/model.py.txt)
[thin_wrapper.py.txt](https://github.com/pytorch/pytorch/files/14750200/thin_wrapper.py.txt)

### Versions

pytorch as of 3/23

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

**Comment by mikekgfb** [comment]
cc: @malfet @desertfire @Chillee @ianbarber @orionr

**Comment by mikekgfb** [comment]
Here are the logits produced for PyTorch eager, and by compiled model, for prefill.
(ref-logits referring to the eager reference numbers)

[mikekg@devgpu018.ftw1 ~/gpt_fast]$ pastry <~/ref-logits
P1200556088: https://www.internalfb.com/intern/paste/P1200556088/
[mikekg@devgpu018.ftw1 ~/gpt_fast]$ pastry </tmp/logits
P1200556173: https://www.internalfb.com/intern/paste/P1200556173/

**Comment by desertfire** [comment]
Likely related to https://github.com/pytorch/pytorch/issues/120424

**Comment by shunting314** [comment]
From triage review: the issue is already solved.
